DB設計（ER図作成）

shota-ino opened the pull request:
ER図
https://gyazo.com/5890e228a42dce79dd717ed5761fe588

#What
データベースの設計、4つのテーブルを作成しそれぞれのアソシエーションを記載
※ActiveHashについてがまだ理解不足で、外部キー参照型に変更した方が良いのかが不明です。
#Why
今後のアプリ作成の根幹となる為
※参考カリキュラムをどのタイミングで読めば良いのか不明の為、一旦提出させて頂きました。

shota-ino committed:
DB設計（ER図作成）

NaturalRadio commented on README.md:
```diff
@@ -1,24 +1,75 @@
+| ---------------- | ------ | -------------------------------------- |
+| nickname         | string | null: false                            |
+| email            | string | null: false, unique: true, index: true |
+| password         | string | null: false                            |
```
```
| password         | string | null: false                            |
```
こちらのカラムはdeviseのgemを使用しての実装になるため、deviseが予め用意している
encrypted_passwordというカラムを用意してあげましょう。

NaturalRadio commented on README.md:
```diff
@@ -1,24 +1,75 @@
-* Services (job queues, cache servers, search engines, etc.)
+| Column          | Type       | Options                                |
+| --------------- | ---------- | -------------------------------------- |
+| item_image      |            | activestorege                          |
```
```
| item_image      |            | activestorege                          |
```
こちらのカラムですが、今回の実装では商品出品の画像投稿において「active_storage」を使用するため、設計の段階から削除しておきましょう。
（理由はactive_storage導入時に自動でテーブルなどが生成されるからです）

NaturalRadio commented on README.md:
```diff
@@ -1,24 +1,75 @@
+| item_image      |            | activestorege                          |
+| name            | string     | null: false, unique: true, index: true |
+| introduction    | text       | null: false                            |
+| category        | string     | null: false                            |
+| item_condition  | string     | null: false                            |
+| postage_payer   | string     | null: false                            |
+| prefecture_code | integer    | null: false                            |
+| preparation_day | string     | null: false                            |
```
**商品の状態**・**配送料の負担**・**発送元の地域**・**発送までの日数**・**カテゴリー**
これらのカラムはActiveHashで実装の予定のため、カラム名末尾に`_id`をつけておくと、この先実装を進めていく中でエラー発生を避けることができます。

NaturalRadio commented on README.md:
```diff
@@ -1,24 +1,75 @@
+
+| Column    | Type       | Options                        |
+| --------- | ---------- | ------------------------------ |
+| item_id   | references | null: false, foreign_key: true |
+| user_id   | references | null: false, foreign_key: true |
```
references 型で記述する際は、カラム名の_idは不要となります。
もしカラム名で_idを使用する予定であれば、integer型を使用した方が良いでしょう。

NaturalRadio commented on README.md:
```diff
@@ -1,24 +1,75 @@
+| Column          | Type       | Options                        |
+| --------------- | ---------- | ------------------------------ |
+| post_code       | integer(7) | null: false                    |
+| prefecture_code | integer    | null: false                    |
```
```
| prefecture_code | integer    | null: false                    |
```
こちらのカラムはActiveHashで実装の予定のため、カラム名末尾に_idをつけておくと、この先実装を進めていく中でエラー発生を避けることができます。

NaturalRadio reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

shota-ino commented:
レビュー頂きありがとうございます。
全体の大枠は大きく間違ってはいないという認識（また、必要なテーブルとエンティティは出せている）でよろしいでしょうか？
ご指摘頂きました点、変更しました。
再度ご確認をお願い申し上げます。

shota-ino committed:
DB設計（修正1回目）

devise導入に向けencrypted_passwordカラムの設定
active_storege導入の為、item_imageカラムの削除
ActiveHash導入の為、_idの追記
references型使用時のカラム名変更

chirikuzu commented on README.md:
```diff
@@ -1,24 +1,74 @@
-* Services (job queues, cache servers, search engines, etc.)
+| Column             | Type       | Options                                |
+| ------------------ | ---------- | -------------------------------------- |
+| name               | string     | null: false, unique: true, index: true |
```
```
| name               | string     | null: false, unique: true, index: true |
```
こちら商品名に一意性制約をかけてしまうと、同じ名前の商品を出品することができなくなってしまいます。
そのため、こちらは一意性制約は外しておきましょう。

chirikuzu commented on README.md:
```diff
@@ -1,24 +1,74 @@
+
+| Column             | Type       | Options                        |
+| ------------------ | ---------- | ------------------------------ |
+| post_code          | integer(7) | null: false                    |
```
```
| post_code          | integer(7) | null: false                    |
```
郵便番号はハイフンを必要とすることも考慮してstring型で実装しましょう。

chirikuzu commented on README.md:
```diff
@@ -1,24 +1,74 @@
+| post_code          | integer(7) | null: false                    |
+| prefecture_code_id | integer    | null: false                    |
+| city               | string     | null: false                    |
+| house_number       | integer    | null: false, unique: true      |
```
```
| house_number       | integer    | null: false, unique: true      |
```
番地にはハイフンや丁目といった文字列が含まれることも考慮してstring型で設計しましょう。

chirikuzu commented on README.md:
```diff
@@ -1,24 +1,74 @@
+| city               | string     | null: false                    |
+| house_number       | integer    | null: false, unique: true      |
+| building_name      | string     |                                |
+| phone_number       | integer    | null: false, unique: true      |
```
```
| phone_number       | integer    | null: false, unique: true      |
```
電話番号はstring型で実装しましょう。
integer型として保存すると整数型として保存されるため、先頭の0がDB上で省略された状態で保存されてしまうためです。

また、今回の実装では配送先住所は購入のたびに入力する実装を行うため、一意性制約をつけてしまうと同じユーザーが2回以上購入を行うことができなくなってしまいます。
こちらは、一意性制約を外しておきましょう。

chirikuzu reviewed (commented):
レビュー依頼ありがとうございます。
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いします。

shota-ino committed:
DB設計（修正2回目）

型の変更
一意性の削除

shota-ino commented:
再度レビューありがとうございます。
ご指摘を受け、今回のフリマアプリの概要（購入時に都度、配送先入力）を改めて認識いたしました。
再修正いたしましたので、ご確認のほど宜しくお願いします。

masahirokurauchi reviewed (commented):
レビュー依頼ありがとうございます！
LGTMとさせていただきます！実装お疲れ様でした！
引き続き実装頑張っていきましょう！

shota-ino commented:
ありがとうございます。引き続き進めて行きたいと思います。
何度もお手数をおかけしました。
